File: app/src/main/java/com/weiwa/ljl/weiwa/ImageFragment.java
package com.weiwa.ljl.weiwa;


import android.content.Intent;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.net.Uri;
import android.os.AsyncTask;
import android.os.Bundle;
import android.os.Environment;
import android.support.annotation.NonNull;
import android.app.Fragment;
import android.support.v4.view.ViewPager;
import android.support.v7.widget.RecyclerView;
import android.util.DisplayMetrics;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageButton;
import android.widget.ImageView;
import android.widget.TextView;
import android.widget.Toast;

import com.bumptech.glide.Glide;
import com.bumptech.glide.load.engine.DiskCacheStrategy;
import com.bumptech.glide.request.animation.GlideAnimation;
import com.bumptech.glide.request.target.GlideDrawableImageViewTarget;
import com.bumptech.glide.request.target.SimpleTarget;

import java.io.File;
import java.io.FileInputStream;
import java.io.FileNotFoundException;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.net.HttpURLConnection;
import java.net.MalformedURLException;
import java.net.URL;
import java.util.concurrent.ExecutionException;

import javax.microedition.khronos.opengles.GL11;

import uk.co.senab.photoview.PhotoViewAttacher;


/**
 * A simple {@link Fragment} subclass.
 */
public class ImageFragment extends Fragment {

    private View mView;
    private File downloadFile;
    private File downloadCategory = new File(Environment.getExternalStorageDirectory().getAbsolutePath()+"/Weiwa");
    private TextView mProgressText;
    private TextView mDownloadText;
    private ImageButton mShareButton;
    private ImageViewPager mViewPager;
    private Uri[] uris;
    private int currentIndex = 0;
    private int width;
    private int height;

    public ImageFragment() {
        // Required empty public constructor
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        if(!downloadCategory.exists()){
            downloadCategory.mkdir();
        }
        if (getArguments() != null) {
            //init Uris
            uris = (Uri[]) (getArguments().getParcelableArray("Uris"));
            currentIndex = getArguments().getInt("Index");
        }
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        // Inflate the layout for this fragment
        DisplayMetrics displayMetrics = new DisplayMetrics();
        getActivity().getWindowManager().getDefaultDisplay().getMetrics(displayMetrics);
        width = displayMetrics.widthPixels;
        height = displayMetrics.heightPixels;
        mView = inflater.inflate(R.layout.fragment_image, container, false);
        mShareButton = (ImageButton) mView.findViewById(R.id.button_share);
        mShareButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                try {
                    share();
                } catch (ExecutionException e) {
                    e.printStackTrace();
                } catch (InterruptedException e) {
                    e.printStackTrace();
                }
            }
        });
        mProgressText = (TextView) mView.findViewById(R.id.progress_text);
        mProgressText.setText(currentIndex+1+"/"+uris.length);
        mDownloadText = (TextView) mView.findViewById(R.id.download_text) ;
        mViewPager = (ImageViewPager) mView.findViewById(R.id.guide_viewpager);
        mViewPager.setMinimumWidth(RecyclerView.LayoutParams.MATCH_PARENT);
        mViewPager.init(getActivity(), uris, new ImageViewPager.onImageDownladingListener() {
            @Override
            public void onDownload(Uri uri, ImageView imageView) {
                try {
                    LoadGlideImage(uri,imageView);
                } catch (MalformedURLException e) {
                    e.printStackTrace();
                }
            }
        },currentIndex);
        mViewPager.addOnPageChangeListener(new ViewPager.OnPageChangeListener() {
            @Override
            public void onPageScrolled(int position, float positionOffset, int positionOffsetPixels) {
                setProgressText((position+1)+"/"+uris.length);
            }

            @Override
            public void onPageSelected(int position) {

            }

            @Override
            public void onPageScrollStateChanged(int state) {

            }
        });
        return mView;
    }

    public void setProgressText(String content){
        mProgressText.setText(content);
    }

    public void LoadGlideImage(final Uri targetUri, ImageView imageView) throws MalformedURLException {
        if(isGif(targetUri)){
            GlideDrawableImageViewTarget imageViewTarget = new GlideDrawableImageViewTarget(imageView);
            Glide.with(this).load(targetUri).downloadOnly(new SimpleTarget<File>() {
                @Override
                public void onResourceReady(File resource, GlideAnimation<? super File> glideAnimation) {
                    try {
                        FileInputStream inputStream = new FileInputStream(resource);
                        downloadFile = new File(downloadCategory, WeiwaApplication.getFileName(targetUri)+".gif");
                        FileOutputStream outputStream = new FileOutputStream(downloadFile);
                        int bytesRead = 0;
                        byte[] buffer = new byte[8192];
                        while ((bytesRead = inputStream.read(buffer, 0, 8192)) != -1) {
                            outputStream.write(buffer,0,bytesRead);
                        }
                        inputStream.close();
                        outputStream.close();
                    } catch (FileNotFoundException e) {
                        e.printStackTrace();
                    }catch(IOException e){
                        e.printStackTrace();
                    }
                }
            });
            Glide.with(this).load(targetUri).diskCacheStrategy(DiskCacheStrategy.SOURCE).crossFade().into(imageViewTarget);
        }else {
            BitmapDownloader bitmapDownloader = new BitmapDownloader();
            //current item position is used to display progress of item downloading
            bitmapDownloader.execute(new Object[]{(new URL(targetUri.toString())),imageView,mViewPager.getCurrentItem()});
        }
    }

    private void share() throws ExecutionException, InterruptedException {
        Intent sendIntent = new Intent(Intent.ACTION_SEND);
        //set intent type
        sendIntent.setType("image/*");
        File shareFile = new File(WeiwaApplication.CacheCategory+"/"+WeiwaApplication.getFileName(uris[mViewPager.getCurrentItem()]));
        Uri targetUri = Uri.fromFile(shareFile);
        if (targetUri != null) {
            sendIntent.putExtra(Intent.EXTRA_STREAM, targetUri);
        }
        sendIntent.putExtra(Intent.EXTRA_SUBJECT, "分享");
        sendIntent.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
        startActivity(Intent.createChooser(sendIntent, "share"));
    }

    @NonNull
    private Boolean isGif(Uri uri){
        if(uri.toString().toLowerCase().endsWith("gif")){
            return true;
        }else{
            return false;
        }
    }

    class BitmapDownloader extends AsyncTask<Object,Object,File> {
        private ImageView imageView;
        @Override
        protected File doInBackground(Object... params) {
            Bitmap bitmap;
            try {
                //获得连接
                URL url = (URL) params[0];
                imageView = (ImageView)params[1];
                int position = (int) params[2];
                HttpURLConnection conn = (HttpURLConnection) url.openConnection();
                //设置超时时间为6000毫秒，conn.setConnectionTiem(0);表示没有时间限制
                conn.setConnectTimeout(6000);
                //连接设置获得数据流
                conn.setDoInput(true);
                //不使用缓存
                conn.setUseCaches(false);
                //这句可有可无，没有影响
                //conn.connect();
                //得到数据流
                InputStream is = conn.getInputStream();
                downloadFile = new File(downloadCategory, WeiwaApplication.getFileName(url));
                OutputStream os = new FileOutputStream(downloadFile);
                int bytesRead = 0;
                int count = conn.getContentLength();
                byte[] buffer = new byte[count];
                float progress = 0.0f;
                int offset = 0;
                int read = 8192;
                if(read>count){
                    read = count;
                }
                while ((bytesRead = is.read(buffer, offset, read)) != -1) {
                    os.write(buffer,offset,bytesRead);
                    offset += bytesRead;
                    if(is.available()!=0) {
                        progress = (float)offset / count;
                    }else{
                        progress = (float)offset / count;
                    }
                    if(count-offset<=read){
                        read = count-offset;
                    }
                    if(WeiwaApplication.getFileName(url).equals(WeiwaApplication.getFileName(uris[position]))){
                        publishProgress(progress);
                    }
                }
                os.close();
                is.close();
                return downloadFile;
            } catch (Exception ex) {
                ex.printStackTrace();
            }
            return null;
        }

        @Override
        protected void onProgressUpdate(Object... values) {
            int display_progress = (int)((float)values[0]*100);
            mDownloadText.setVisibility(View.VISIBLE);
            mDownloadText.setText(display_progress+"%");
            if(display_progress==1){
                mDownloadText.setText("0");
                mDownloadText.setVisibility(View.INVISIBLE);
            }
        }

        @Override
        protected void onPostExecute(File result) {
            //GL11.GL_MAX_TEXTURE_SIZE
            if(result==null){
                Toast.makeText(ImageFragment.this.getActivity(),"文件下载失败",Toast.LENGTH_SHORT).show();
                return;
            }
            imageView.setImageURI(Uri.fromFile(result));
            if((float)(imageView.getDrawable().getIntrinsicHeight()/imageView.getDrawable().getIntrinsicWidth())>2.0f){
                //int newHeight = (int) ( result.getHeight() * (512.0 / result.getWidth()) );
                //Bitmap putImage = Bitmap.createScaledBitmap(result, 512, newHeight, true);
                //int height = putImage.getHeight();
                //imageView.setImageBitmap(result);
                //imageView.invalidate();
                //int w1 = result.getWidth();
                PhotoViewAttacher attacher = new PhotoViewAttacher(imageView);
                float ratio = (float)imageView.getWidth()/(float)(attacher.getImageView().getDrawable().getIntrinsicWidth());
                if(ratio<4 && ratio>2){
                    ratio = ratio/2;
                }
                attacher.setScaleLevels(ratio/2,ratio,ratio*2);
                attacher.setScale(ratio,ratio,ratio,true);
            }else {
                PhotoViewAttacher attacher = new PhotoViewAttacher(imageView);
            }
        }
    }
}

File: app/src/main/java/com/weiwa/ljl/weiwa/fragment/ImageFragment.kt
package com.weiwa.ljl.weiwa.fragment


import android.app.Fragment
import android.content.Intent
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import android.net.Uri
import android.os.AsyncTask
import android.os.Bundle
import android.os.Environment
import android.support.v4.view.ViewPager
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageButton
import android.widget.ImageView
import android.widget.TextView
import android.widget.Toast
import com.bumptech.glide.Glide
import com.bumptech.glide.load.engine.DiskCacheStrategy
import com.bumptech.glide.request.animation.GlideAnimation
import com.bumptech.glide.request.target.GlideDrawableImageViewTarget
import com.bumptech.glide.request.target.SimpleTarget
import com.diegocarloslima.byakugallery.lib.TileBitmapDrawable
import com.wang.avi.AVLoadingIndicatorView
import com.weiwa.ljl.weiwa.R
import com.weiwa.ljl.weiwa.WeiwaApplication
import com.weiwa.ljl.weiwa.network.NetworkBroadcastReceiver
import com.weiwa.ljl.weiwa.view.ImageViewPager
import com.weiwa.ljl.weiwa.view.TouchImageView
import uk.co.senab.photoview.PhotoViewAttacher
import java.io.*
import java.net.HttpURLConnection
import java.net.MalformedURLException
import java.net.URL
import java.util.concurrent.ExecutionException


/**
 * A simple [Fragment] subclass.
 */
class ImageFragment : Fragment() {

    private var mView: View? = null
    private var downloadFile: File? = null
    private val downloadCategory = File(Environment.getExternalStorageDirectory().absolutePath + "/Weiwa")
    private var mProgressText: TextView? = null
    private var mDownloadText: TextView? = null
    private var mShareButton: ImageButton? = null
    private var mViewPager: ImageViewPager? = null
    private var mAvi: AVLoadingIndicatorView? = null
    private var uris: Array<Uri>? = null
    private var currentIndex = 0

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        if (!downloadCategory.exists()) {
            downloadCategory.mkdir()
        }
        if (arguments != null) {
            //init Uris
            uris = arguments.getParcelableArray("Uris") as Array<Uri>
            currentIndex = arguments.getInt("Index")
        }
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        // Inflate the layout for this fragment
        mView = inflater.inflate(R.layout.fragment_image, container, false)
        mAvi = mView!!.findViewById(R.id.avi) as AVLoadingIndicatorView
        mShareButton = mView!!.findViewById(R.id.button_share) as ImageButton
        mShareButton!!.setOnClickListener {
            try {
                share()
            } catch (e: ExecutionException) {
                e.printStackTrace()
            } catch (e: InterruptedException) {
                e.printStackTrace()
            }
        }
        mProgressText = mView!!.findViewById(R.id.progress_text) as TextView
        mProgressText!!.text = (currentIndex + 1).toString() + "/" + uris!!.size
        mDownloadText = mView!!.findViewById(R.id.download_text) as TextView
        mViewPager = mView!!.findViewById(R.id.guide_viewpager) as ImageViewPager
        mViewPager!!.minimumWidth = RecyclerView.LayoutParams.MATCH_PARENT
        mViewPager!!.init(activity, uris!!, object : ImageViewPager.onImageDownladingListener {
            override fun onDownload(uri: Uri, imageView: ImageView) {
                try {
                    LoadGlideImage(uri, imageView)
                } catch (e: MalformedURLException) {
                    e.printStackTrace()
                }
            }
        }, currentIndex)
        mViewPager!!.addOnPageChangeListener(object : ViewPager.OnPageChangeListener {
            override fun onPageScrolled(position: Int, positionOffset: Float, positionOffsetPixels: Int) {
                setProgressText((position + 1).toString() + "/" + uris!!.size)
            }

            override fun onPageSelected(position: Int) {

            }

            override fun onPageScrollStateChanged(state: Int) {

            }
        })
        return mView
    }

    fun setProgressText(content: String) {
        mProgressText!!.text = content
    }

    @Throws(MalformedURLException::class)
    fun LoadGlideImage(targetUri: Uri, imageView: ImageView) {
        mAvi!!.show()
        if (isGif(targetUri)) {
            val imageViewTarget = GlideDrawableImageViewTarget(imageView)
            Glide.with(this).load(targetUri).downloadOnly<SimpleTarget<File>>(object : SimpleTarget<File>() {
                override fun onResourceReady(resource: File, glideAnimation: GlideAnimation<in File>) {
                    try {
                        val inputStream = FileInputStream(resource)
                        downloadFile = File(downloadCategory, WeiwaApplication.getFileName(targetUri) + ".gif")
                        val outputStream = FileOutputStream(downloadFile!!)
                        var bytesRead = 0
                        val buffer = ByteArray(8192)
                        while ({ bytesRead = inputStream.read(buffer, 0, 8192); bytesRead }() != -1) {
                            outputStream.write(buffer, 0, bytesRead)
                        }
                        inputStream.close()
                        outputStream.close()
                        mAvi!!.hide()
                    } catch (e: FileNotFoundException) {
                        mAvi!!.hide()
                        e.printStackTrace()
                    } catch (e: IOException) {
                        mAvi!!.hide()
                        e.printStackTrace()
                    }

                }
            })
            Glide.with(this).load(targetUri).diskCacheStrategy(DiskCacheStrategy.SOURCE).fitCenter().into(imageViewTarget)
        } else {
            if (!NetworkBroadcastReceiver.getWifiState()) {
                imageView.setImageDrawable(this.activity.getDrawable(R.drawable.ic_wait_for_download))
                imageView.setOnClickListener {
                    val bitmapDownloader = BitmapDownloader()
                    //current item position is used to display progress of item downloading
                    bitmapDownloader.execute(*arrayOf(URL(targetUri.toString()), imageView, mViewPager!!.currentItem))
                }
                return
            }
            val bitmapDownloader = BitmapDownloader()
            //current item position is used to display progress of item downloading
            bitmapDownloader.execute(*arrayOf(URL(targetUri.toString()), imageView, mViewPager!!.currentItem))
        }
    }

    @Throws(ExecutionException::class, InterruptedException::class)
    private fun share() {
        val shareFile: File
        val sendIntent = Intent(Intent.ACTION_SEND)
        //set intent type
        sendIntent.type = "image/*"
        if (isGif(uris!![mViewPager!!.currentItem])) {
            shareFile = File(WeiwaApplication.CacheCategory.toString() + "/" + WeiwaApplication.getFileName(uris!![mViewPager!!.currentItem]) + ".gif")
        } else {
            shareFile = File(WeiwaApplication.CacheCategory.toString() + "/" + WeiwaApplication.getFileName(uris!![mViewPager!!.currentItem]))
        }
        val targetUri = Uri.fromFile(shareFile)
        if (targetUri != null) {
            sendIntent.putExtra(Intent.EXTRA_STREAM, targetUri)
        }
        sendIntent.putExtra(Intent.EXTRA_SUBJECT, "分享")
        sendIntent.flags = Intent.FLAG_ACTIVITY_NEW_TASK
        startActivity(Intent.createChooser(sendIntent, "share"))
    }

    private fun isGif(uri: Uri): Boolean {
        return uri.toString().toLowerCase().endsWith("gif")
    }

    internal inner class BitmapDownloader : AsyncTask<Any, Any, File>() {
        private var imageView: TouchImageView? = null
        private var url: URL? = null
        private var position: Int = 0
        override fun doInBackground(vararg params: Any): File? {
            val bitmap: Bitmap
            try {
                //获得连接
                url = params[0] as URL
                imageView = params[1] as TouchImageView
                position = params[2] as Int
                val conn = url!!.openConnection() as HttpURLConnection
                //设置超时时间为6000毫秒，conn.setConnectionTiem(0);表示没有时间限制
                conn.connectTimeout = 6000
                //连接设置获得数据流
                conn.doInput = true
                //不使用缓存
                conn.useCaches = false
                //得到数据流
                val `is` = conn.inputStream
                downloadFile = File(downloadCategory, WeiwaApplication.getFileName(url!!))
                val os = FileOutputStream(downloadFile!!)
                var bytesRead = 0
                val count = conn.contentLength
                val buffer = ByteArray(count)
                var progress = 0.0f
                var offset = 0
                var read = 8192
                if (read > count) {
                    read = count
                }
                while ({ bytesRead = `is`.read(buffer, offset, read); bytesRead }() != -1) {
                    os.write(buffer, offset, bytesRead)
                    offset += bytesRead
                    if (`is`.available() != 0) {
                        progress = offset.toFloat() / count
                    } else {
                        progress = offset.toFloat() / count
                    }
                    if (count - offset <= read) {
                        read = count - offset
                    }
                    if (WeiwaApplication.getFileName(url!!) == WeiwaApplication.getFileName(uris!![position])) {
                        publishProgress(progress)
                    }
                }
                os.close()
                `is`.close()
                return downloadFile
            } catch (ex: Exception) {
                ex.printStackTrace()
            }
            return null
        }

        override fun onProgressUpdate(vararg values: Any) {
            val display_progress = (values[0] as Float * 100).toInt()
            mDownloadText!!.visibility = View.VISIBLE
            mDownloadText!!.text = display_progress.toString() + "%"
            if (display_progress == 1) {
                mDownloadText!!.text = "0"
                mDownloadText!!.visibility = View.INVISIBLE
            }
        }

        override fun onPostExecute(result: File?) {
            //GL11.GL_MAX_TEXTURE_SIZE
            if (result == null) {
                Toast.makeText(this@ImageFragment.activity, "文件下载失败", Toast.LENGTH_SHORT).show()
                mAvi!!.hide()
                return
            }
            if (imageView == null) {
                mAvi!!.hide()
                return
            }
            imageView!!.setImageURI(Uri.fromFile(result))
            if ((imageView!!.drawable.intrinsicHeight / imageView!!.drawable.intrinsicWidth).toFloat() > 2.0f) {
                val newBitmap = BitmapFactory.decodeFile(result.absolutePath)
                val newWidth = imageView!!.width
                val ratio = newWidth.toFloat() / newBitmap.width.toFloat()
                imageView!!.setMaxScale(ratio)
                TileBitmapDrawable.attachTileBitmapDrawable(imageView, result.absolutePath, null, null)
                if (WeiwaApplication.getFileName(url!!) == WeiwaApplication.getFileName(uris!![position])) {
                    imageView!!.performDoubleTap()
                }
            } else {
                val attacher = PhotoViewAttacher(imageView!!)
            }
            mAvi!!.hide()
        }
    }
}// Required empty public constructor
